### PR TITLE
scons: use github instead of sourceforge

### PR DESCRIPTION
**Pull request by AndersonTorres**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/tools/build-managers/scons/4.5.2.nix [hunk 1]
```diff
@@ -21,9 +25,13 @@ python3.pkgs.buildPythonApplication {
...
 
+  preConfigure = ''
+    python scripts/scons.py
+  '';
+
   postInstall = ''
-    mkdir -p "$out/share/man/man1"
-    mv "$out/"*.1 "$out/share/man/man1/"
+    mkdir -p "$man/share/man/man1"
+    mv "$out/"*.1 "$man/share/man/man1/"
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/scons/4.5.2.nix:34]:
```suggestion
    installManPage $out/*.1
```

File: pkgs/development/tools/build-managers/scons/4.1.0.nix [hunk 2]
```diff
@@ -1,25 +1,32 @@
...
+
+  preConfigure = ''
+    python scripts/scons.py
   '';
 
   postInstall = ''
-    mkdir -p "$out/share/man/man1"
-    mv "$out/"*.1 "$out/share/man/man1/"
+    mkdir -pv "$man/share/man/man1"
+    mv -v "$out/"*.1 "$man/share/man/man1/"
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/scons/4.1.0.nix:29]:
```suggestion
    installManPage $out/*.1
```

File: pkgs/development/tools/build-managers/scons/3.1.2.nix [hunk 3]
```diff
@@ -36,3 +45,4 @@ python3.pkgs.buildPythonApplication {
     maintainers = with lib.maintainers; [ AndersonTorres ];
   };
 }
+# TODO: patch to get rid of distutils and other deprecations
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/scons/3.1.2.nix:48]: we should report that to upstream. There are thousands of packages having this problem.

Comment (AndersonTorres) [reply pkgs/development/tools/build-managers/scons/3.1.2.nix:48]: I don't believe they will care about an older release from December 2019.
It should be easier to get rid of the softwares that require Scons 3.1.2 instead.

File: pkgs/development/tools/build-managers/scons/4.5.2.nix (hunk 1 above)
Comment (AndersonTorres) [reply pkgs/development/tools/build-managers/scons/4.5.2.nix:34]: I don't believe this is needed.

File: pkgs/development/tools/build-managers/scons/4.1.0.nix [hunk 4]
```diff
@@ -1,25 +1,32 @@
...
 
 let
   pname = "scons";
   version = "4.1.0";
-  src = fetchurl {
-    url = "mirror://sourceforge/scons/scons-${version}.tar.gz";
-    hash = "sha256-ctKNdi4hJnh/Fz49WeCJI5+LL06e8xFNV/ELEgaYXYU=";
+  src = fetchFromGitHub {
+    owner = "Scons";
+    repo = "scons";
```
Comment (matdibu) [inline pkgs/development/tools/build-managers/scons/4.1.0.nix:8]:
```suggestion
    repo = pname;
```

File: pkgs/development/tools/build-managers/scons/4.5.2.nix [hunk 5]
```diff
@@ -1,16 +1,20 @@
...
 
 let
   pname = "scons";
   version = "4.5.2";
-  src = fetchurl {
-    url = "mirror://sourceforge/project/scons/scons/${version}/SCons-${version}.tar.gz";
-    hash = "sha256-ziaqyV01CnmkGSGWsL6sPLJPTMq84BI+so0zcPV28HI=";
+  src = fetchFromGitHub {
+    owner = "Scons";
+    repo = "scons";
```
Comment (matdibu) [inline pkgs/development/tools/build-managers/scons/4.5.2.nix:8]:
```suggestion
    repo = pname;
```

File: pkgs/development/tools/build-managers/scons/3.1.2.nix [hunk 6]
```diff
@@ -1,16 +1,25 @@
...
 
 let
   pname = "scons";
   version = "3.1.2";
-  src = fetchurl {
-    url = "mirror://sourceforge/scons/scons-${version}.tar.gz";
-    hash = "sha256-eAHz9i9lRSjict94C+EMDpM36JdlC2Ldzunzn94T+Ps=";
+  src = fetchFromGitHub {
+    owner = "Scons";
+    repo = "scons";
```
[the remark this answers is not in the document]
Comment (AndersonTorres) [reply pkgs/development/tools/build-managers/scons/3.1.2.nix:8]: NO.

File: pkgs/development/tools/build-managers/scons/4.1.0.nix (hunk 4 above)
Comment (AndersonTorres) [reply pkgs/development/tools/build-managers/scons/4.1.0.nix:8]: No.

File: pkgs/development/tools/build-managers/scons/4.5.2.nix (hunk 5 above)
Comment (AndersonTorres) [reply pkgs/development/tools/build-managers/scons/4.5.2.nix:8]: What do you suggest to `owner` macro-substitution?

File: pkgs/development/tools/build-managers/scons/4.5.2.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/tools/build-managers/scons/4.5.2.nix:34]: Oh, yeah, that would only work if the man page wasn't already installed.

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/development/tools/build-managers/scons/3.1.2.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (matdibu) [reply pkgs/development/tools/build-managers/scons/3.1.2.nix:8]: The suggestion was based on https://github.com/NixOS/nixpkgs/tree/master/pkgs#package-naming, more specifically:

> There are a few naming guidelines:
> - The pname attribute should be identical to the upstream package name.

and [5.3k other files in the repo](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22repo+%3D+pname%22&type=code).

I'm a bit new so I don't know if the consensus/best practices have changed.

**Comment by matdibu** [comment]
The `repo = pname;` suggestion was based on https://github.com/NixOS/nixpkgs/tree/master/pkgs#package-naming, more specifically:

> There are a few naming guidelines:
> - The pname attribute should be identical to the upstream package name.

and [5.3k other files in the repo](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22repo+%3D+pname%22&type=code).

I'm new to nix and nixpkgs so I don't know if the consensus/best practices have changed.

**Comment by AndersonTorres** [comment]
> The `repo = pname;` suggestion was based on https://github.com/NixOS/nixpkgs/tree/master/pkgs#package-naming, more specifically:
> 
> > There are a few naming guidelines:
> > 
> > * The pname attribute should be identical to the upstream package name.

First, your reasoning is backwards. If `pname` should be determined by the upstream package name, then your suggestion should be `pname = repo` or `pname =src.repo`, not the reverse.
(And no, `a=b` is not `b=a`.)

Second, even under this rule `repo` is identical to `pname`, therefore your "fixup" is redundant.

Third, the text does not say we should write `repo = pname` always when possible.

Fourth, it is not always that the GitHub repo link has the same name of the upstream package name.

An example was the now-defunct `pil`, the Python Imaging Library. Its "upstream package name" was `pil`, however the tarball was named `Imaging-${version}.tgz`. I would like to see how `repo = pname` would help in this case.

Another more modern example is OpenMSX...

Fifth, Nix is not a macro substitution language. Trying to parameterize everything is useless.

> 
> and [5.3k other files in the repo](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22repo+%3D+pname%22&type=code).

5k files in a repo with 31k files? This is not a huge evidence.

Even if we pick only [`repo =` appearances](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22repo+%3D%22&type=code), we have 14k, way more than double. Therefore, the number of  appearances of `repo =  pname` is smaller than the rest.
 
Also, those appearances of `= pname` serve to nothing besides creating useless and harmful in-code dependencies.

**Review by pbsds — APPROVED** [review]
@matdibu, welcome to nixpkgs and please don't feel dejected by the tone of the feedback. Your contribution in #272966 was stellar.

---

Diff LGTM, can't build or test locally due to all the rebuilds however

**Pull request merged**